pyscraper/base_resolver.py

```diff
@@ -0,0 +1,169 @@
+        self.membertopersonmap[mship["id"]] = mship['person_id']
+        self.persontomembermap.setdefault(mship['person_id'], []).append(mship["id"])
+
+        if self.members.get(mship["id"]):
+            raise Exception, "Repeated identifier %s in members JSON file" % mship["id"]
+        self.members[mship["id"]] = mship
+
+        if 'end_date' not in mship:
+            mship['end_date'] = '9999-12-31'
+
+        family_name = mship['name']["family_name"]
+        given_name = mship['name']["given_name"]
+
+        # index by "Firstname Lastname" for quick lookup ...
+        compoundname = self.member_full_name(mship['id'])
+        self.fullnames.setdefault(compoundname, []).append(mship)
+
+        # add in names without the middle initial
+        fnnomidinitial = re.findall('^(\S*)\s\S$', given_name)
+        if fnnomidinitial:
+            compoundname = fnnomidinitial[0] + " " + family_name
+            self.fullnames.setdefault(compoundname, []).append(mship)
+
+        if self.import_organization_id != 'house-of-commons':
+            # ... and by first initial, lastname
+            if given_name:
+                compoundname = given_name[0] + " " + family_name
+                self.fullnames.setdefault(compoundname, []).append(mship)
+
+        # ... and also by "Lastname"
+        self.lastnames.setdefault(family_name, []).append(mship)
+
+        # ... and by constituency
+        mship['constituency'] = posts[mship['post_id']]['area']['name']
+        consids = self.constoidmap[mship['constituency']]
+        consid = None
+        # find the constituency id for this person
+        mship_start_date = len(mship['start_date'])==4 and ('%s-01-01' % mship['start_date']) or mship['start_date']
+        mship_end_date = len(mship['end_date'])==4 and ('%s-12-31' % mship['end_date']) or mship['end_date']
+        for cons in consids:
+            if (cons['start_date'] <= mship_start_date and
+                mship_start_date <= mship_end_date and
+                mship_end_date <= cons['end_date']):
+                if consid and consid != cons['id']:
+                    raise Exception, "Two constituency ids %s %s overlap with MP %s" % (consid, cons['id'], mship['id'])
+                consid = cons['id']
+        if not consid:
+            raise Exception, "Constituency '%s' not found" % mship["constituency"]
+        # check name in members file is same as default in cons file
+        backformed_cons = self.considtonamemap[consid]
+        if backformed_cons != mship["constituency"]:
+            raise Exception, "Constituency '%s' in members file differs from first constituency '%s' listed in cons file" % (mship["constituency"], backformed_cons)
+
+        # check first date ranges don't overlap, MPs only
+        if self.import_organization_id == 'house-of-commons':
+            for curattr in self.considtomembermap.get(consid, []):
+                if curattr['end_date'] < '1997-05-01': continue
```

Comment:
I wasn't sure what this was doing :( Thinking about it some more, I think it's because we can only guarantee that memberships don't overlap for non-historic MP memberships (indeed, far enough back we correctly have multiple MPs per constituency), they quite often have same start/end dates previously. I'll add a comment.